homeworks/solutions/torantie/4-param/src/halfedge.js

```diff
@@ -0,0 +1,679 @@
+      for (let j = 0; j < n; j++) {
+        bcycle[j].next = bcycle[(j+n-1)%n]
+        bcycle[j].prev = bcycle[(j+1)%n]
+        hasTwin.set(bcycle[j], true)
+        hasTwin.set(bcycle[j].twin, true)
+      }
+
+    }
+
+    // reset indices
+    let index = 0
+    this.vertices.forEach(v => { v.idx = index++ })
+    index = 0
+    this.edges.forEach(e => { e.idx = index++ })
+    index = 0
+    this.faces.forEach(f => { f.idx = index++ })
+    index = 0
+    this.halfedges.forEach(h => { h.idx = index++ })
+
+  }
+
+
+  /**
+   * flatten computes the UV coordinates of the given triangle mesh.
+   *
+   * This implementation reuiqres the mesh contains at least one
+   * boundary loop.
+   *
+   * @param {string} boundaryType 'disk', or 'rect'
+   * @param {string} laplaceWeight 'uniform', or 'cotan'
+   */
+  flatten(boundaryType, laplaceWeight) {
+    // TODO: implement the Tutte's barycentric embedding algorithm.
+    //
+    // Hint:
+    // 1. check if the mesh contains a boundary
+    let boundaryHalfEdgesLength = this.boundaries.length
+    let numberOfVertices = this.vertices.length
+    let u = DenseMatrix.zeros(numberOfVertices, 1)
+    let v = DenseMatrix.zeros(numberOfVertices, 1)
+
+    let boundaryVerticesCache =new Array(numberOfVertices)
+    let boundaryVertices = []
+
+    if(boundaryHalfEdgesLength == 0)
+      return
+
+    this.boundaries.forEach((face, i) =>{
+      face.vertices((vertex) =>{
+        if(boundaryVerticesCache[vertex.idx] === undefined){
+          boundaryVerticesCache[vertex.idx] = vertex
+          boundaryVertices.push(vertex)
+        }
+      })
+    })
+
+    // 2. compute boundary uv coordinates depending on the boundary type
+    switch (boundaryType) {
+      case "disk":
+        this.calculateDiskBoundary(boundaryVertices, u, v);
+        break;
+
+      case "rect":
+        this.calculateRectBoundary(boundaryVertices, u, v);
+        break;
+    }
+
+    // 3. compute matrix depending on the laplacian weight type
+    let L = this.laplaceMatrix(laplaceWeight,boundaryVerticesCache)
+
+    let luDecomposition = L.lu()
+
+    // 4. solve linear equation and assing computed uv as vertex uv
+    let solvedSquareU= luDecomposition.solveSquare(u)
+    let solvedSquareV= luDecomposition.solveSquare(v)
+
+    for(let index = 0; index<this.vertices.length;index++)
+    {
+      this.updateUVPosition(index, solvedSquareU, solvedSquareV, boundaryVerticesCache);
+    }
+  }
+
+  calculateDiskBoundary(boundaryVertices, u, v) {
+    const r = 0.5 // disk radius
+    boundaryVertices.forEach((vertex, i) => {
+      if (vertex !== null) {
+        let x = (r * Math.cos(2 * Math.PI * i / boundaryVertices.length)) + 0.5
+        let y = (r * Math.sin(2 * Math.PI * i / boundaryVertices.length)) + 0.5
+        vertex.uv.x = x
+        vertex.uv.y = y
+        u.set(x, vertex.idx, 0)
+        v.set(y, vertex.idx, 0)
+      }
+    })
+  }
+
+  calculateRectBoundary(boundaryVertices, u, v) {
```

Comment:
Interesting. I was thinking if you might come up with some nicer implementation (comparing to how I quickly wrote in the reference solution) for handling rectangle boundary vertex positions, but overall the current approach works fine :)